probe/elasticsearch.go

```diff
@@ -0,0 +1,562 @@
+		"holds about 500 pages of text."
+)
+
+const millisecondInMinute = 60_000
+
+type EsDocument struct {
+	Name     string
+	EventTye string
+	Team     string
+	Counter  int
+	Data     string
+}
+
+type EsProbe struct {
+	clusterName   string
+	clusterConfig common.Cluster
+	config        *common.Config
+	client        *elasticsearch7.Client
+
+	consulClient *api.Client
+
+	timeout time.Duration
+
+	updateDiscoveryTicker                 *time.Ticker
+	cleanMetricsTicker                    *time.Ticker
+	executeClusterDurabilityProbingTicker *time.Ticker
+	executeClusterLatencyProbingTicker    *time.Ticker
+	executeNodeProbingTicker              *time.Ticker
+
+	esNodesList         []common.Node
+	allEverKnownEsNodes []string
+
+	controlChan chan bool
+}
+
+func NewEsProbe(clusterName, endpoint string, clusterConfig common.Cluster, config *common.Config, consulClient *api.Client, controlChan chan bool) (EsProbe, error) {
+	var allEverKnownEsNodes []string
+	esNodesList, err := common.DiscoverNodesForService(consulClient, clusterConfig.Name)
+	if err != nil {
+		return EsProbe{}, errors.Wrapf(err, "Impossible to discover ES nodes during bootstrap for cluster %s", clusterName)
+	}
+	allEverKnownEsNodes = common.UpdateEverKnownNodes(allEverKnownEsNodes, esNodesList)
+
+	client, err := initEsClient(clusterConfig.Scheme, endpoint, config.ElasticsearchUser, config.ElasticsearchPassword)
+	if err != nil {
+		return EsProbe{}, errors.Wrapf(err, "Failed to init elasticsearch client for cluster %s", clusterName)
+	}
+
+	return EsProbe{
+		clusterName:   clusterName,
+		clusterConfig: clusterConfig,
+		config:        config,
+		client:        client,
+
+		consulClient: consulClient,
+
+		timeout: config.ProbePeriod - 2*time.Second,
+
+		updateDiscoveryTicker:                 time.NewTicker(config.ConsulPeriod),
+		executeClusterDurabilityProbingTicker: time.NewTicker(config.ProbePeriod),
+		executeClusterLatencyProbingTicker:    time.NewTicker(time.Duration(millisecondInMinute/config.LatencyProbeRatePerMin) * time.Millisecond),
+		executeNodeProbingTicker:              time.NewTicker(config.ProbePeriod),
+		cleanMetricsTicker:                    time.NewTicker(config.CleaningPeriod),
+
+		esNodesList:         esNodesList,
+		allEverKnownEsNodes: allEverKnownEsNodes,
+
+		controlChan: controlChan,
+	}, nil
+}
+
+func (es *EsProbe) PrepareEsProbing() error {
+	// TODO: recreate latency index
+	// Check index available
+	if err := es.createMissingIndex(es.config.ElasticsearchDurabilityIndex); err != nil {
+		return err
+	}
+	if err := es.createMissingIndex(es.config.ElasticsearchLatencyIndex); err != nil {
+		return err
+	}
+
+	// Count docs on durability index and put docs if needed
+	number_of_current_durability_documents, _, err := es.countNumberOfDurabilityDocs()
+	if err != nil {
+		return err
+	}
+
+	if err := es.indexDurabilityDocuments(number_of_current_durability_documents); err != nil {
```

Comment:
It's not easy to understand the purpose of the func from the name; maybe `FillDurabilityBucketWithMissingDocs`